### PR TITLE
feat: introduce `DefaultPropsProvider` for setting default props in React components

### LINKED ISSUE #73: Create a provider that allows customizing default values as well as the theme
## MUI

Open [MUI](https://mui.com/material-ui/getting-started/overview/) and access the `theme` object directly in the console:

Example:
```
components: {
  Link: {
    defaultProps: {
      underline: 'none',
    },
    styleOverrides: {
      root: {
        alignItems: 'center',
        display: 'inline-flex',
        fontWeight: 700,
      },
    },
    variants: [
      {
        props: { color: 'primary' },
        style: f (e),
      }
    ],
  },
}
```

image

## Ant Design - ConfigProvider
https://ant.design/components/config-provider-cn/#header
https://ant.design/components/config-provider-cn/#API

### PR DESCRIPTION
**Pull request by cheton**
https://trendmicro-frontend.github.io/tonic-ui-demo/react/pr-922/

### **User description**
- [x] accordion (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/40a661bffc233c5ff31048fc0ee31d411f2abf92)
  - Accordion
  - AccordionBody
  - AccordionContent
  - AccordionHeader
  - AccordionItem
  - AccordionToggle
  - AccordionToggleIcon
- [x] alert (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/bdd329e4f648b3caa93568deaea962bebac1014e)
  - Alert
  - AlertCloseButton
  - AlertIcon
  - AlertMessage
- [x] badge (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/14c5805d16a1c1d759424c27442901db1a75d3ed)
  - Badge
- [x] ~box~
- [x] button (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/2173038ffd0be52f87006aa7e8c6863e59d39195)
  - Button
  - ButtonBase
  - ButtonGroup
  - ButtonLink
- [x] checkbox (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/9121fa5b3eaf0507317318f3a4074093fa98a8f2)
  - Checkbox
  - CheckboxGroup
- [x] code (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/d0ee6bfac52578db51d4eda7f7ccae117afedf0b)
  - Code
- [x] color-mode (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/d72be6c560818d25538e99c75ed810e3d8135cdf)
  - ColorModeProvider
  - DarkMode
  - InvertedMode
  - LightMode
- [x] color-style (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/f42945e2f4660a7bc6318b93ba874ba92b2cdfdd)
  - ColorStyleProvider
- [x] ~css-baseline~
- [x] date-pickers (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/9e7f816f909c40f9ecb8c68a059fb14f46b4a63b)
  - Calendar
  - DatePicker
- [x] ~default-props~
- [x] ~deprecated~
- [x] divider (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/e1bb7d4133353b4a7bd89702982ef11771703999)
  - Divider
- [x] drawer (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/170e534178306670997fec0a2030e7ac8e3e6f34)
  - Drawer
  - DrawerBody
  - DrawerCloseButton
  - DrawerContainer
  - DrawerContent
  - DrawerFooter
  - DrawerHeader
  - DrawerOverlay
- [x] flex (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/63b72b7b202eeec169b2980c818525c2c24c3019)
  - Flex
- [x] grid (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/ad5d0434764b158ba73f354ec0eb726cdfca5ffd)
  - Grid
- [x] icon (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/6208388ceff55cdb9d6214d2d34331ae54041658)
  - Icon
- [x] image (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/8d8e6f9ad5e558d2424bb5a3bdc09d2889167ad5)
  - Image
- [x] input (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/dcb3c52d842dbcbac9576f7d0e81ce9f3e365811)
  - Input
  - InputAdornment
  - InputBase
  - InputControl
  - InputGroup
  - InputGroupAddon
  - InputGroupAppend
  - InputGroupPrepend
- [x] link (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/d3ea09bc76a8295edd7b2d91357d5397613f3d47)
  - Link
  - LinkButton
- [x] menu (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/c817ff165c05338915f8f1b2303979a30577ed83)
  - MenuButton
  - MenuContent
  - MenuDivider
  - MenuGroup
  - MenuItem
  - Menu
  - MenuList
  - MenuToggleIcon
  - MenuToggle
  - SubmenuContent
  - Submenu
  - SubmenuList
  - SubmenuToggle
- [x] modal (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/6208388ceff55cdb9d6214d2d34331ae54041658)
  - Modal
  - ModalBody
  - ModalCloseButton
  - ModalContainer
  - ModalContent
  - ModalFooter
  - ModalHeader
  - ModalOverlay
- [x] pagination (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/819c3cf78816081eef738b7f7ae6abb5f7f131e2)
  - Pagination
  - PaginationItem
- [x] popover (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/a62b3e6c6d6ec4417130ba1ab75887429f323a04)
  - Popover
  - PopoverArrow
  - PopoverBody
  - PopoverContent
  - PopoverFooter
  - PopoverHeader
  - PopoverTrigger
- [x] ~popper~
- [x] portal (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/fd6471ca147322c46e1ee52cdfdfbfd820c4ff77)
  - Portal
  - PortalManager
- [x] progress (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/09c8e38ff6fafbba125039170bb5dbfe49ff92f2)
  - LinearProgress
- [x] ~provider~
- [x] radio (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/688d98353fd8800c1fbb6b11abeed56dced93c87)
  - Radio
  - RadioGroup
- [x] resize-handle (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/b796ca531590937bd7b02d0d9d4cf5bf38eafffd)
  - ResizeHandle
- [x] scrollbar (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/733568130dbd10609460a1fadfe3268d3c95f6bc)
  - Scrollbar
- [x] search-input (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/299a7023675d6b64e850ac44b31a18df9cb42996)
  - SearchInput
- [x] select (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/e9e7a181c2773b0a17859a2255352c67ce3c86ea)
  - Select
- [x] skeleton (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/b97e9f380c08cd004e7de0ee0ef01bae835e191c)
  - Skeleton
- [x] space (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/35709425a9c8b6b20250b6c2525623b08225abfc)
  - Space
- [x] spinner (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/192601f5ebdf959d89a44936014b4e856e0afc21)
  - Spinner
- [x] stack (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/f7dc2ec5f0be03310e983a6c175da4648112e602)
  - Stack
- [x] switch (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/c22fb25f5e9902dedcf12fd7973a0b40f2890a07)
  - Switch
- [x] table (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/dd564923bf81ab3fa1b82acb534c1e383e963915)
  - Table
  - TableHeader
  - TableHeaderRow
  - TableHeaderCell
  - TableBody
  - TableFooter
  - TableRow
  - TableCell
  - TableScrollbar
- [x] tabs (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/28253d59fa8b0a9da1ae3cbbdf41a24c653d3cd5)
  - Tab
  - Tabs
  - TabList
  - TabPanel
  - TabPanels
- [x] tag (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/3af9c55855695ef0fd5f0a5d0caf8055c83f2786)
  - Tag
  - TagCloseButton
- [x] text (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/3d7f1d8f1ade8eef5755c7f42c40024de7a899c0)
  - Text
- [x] textarea (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/34b840de5c5df50da5789e44fd22d91655ef2872)
  - Textarea
- [x] ~theme~
- [x] toast (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/cbf7b955c564a71fafdd4f0444845d4ded346976)
  - Toast
  - ToastCloseButton
  - ToastContainer
  - ToastController
  - ToastIcon
  - ToastManager
  - ToastMessage
  - ToastTransition
- [x] tooltip (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/45b0b69204f4657f23ec5c7eea6e03d11b9b7170)
  - Tooltip
  - TooltipArrow
  - TooltipContent
  - TooltipTrigger
- [x] transitions (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/ab87b30c986a963c4e024fd9afa16572e8f51f78)
  - Collapse
  - Fade
  - Grow
  - Scale
  - Slide
  - Zoom
- [x] tree (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/50e3a5a5390265a4945b0cf6eab4357836b80616)
  - Tree
  - TreeItem
  - TreeItemContent
  - TreeItemToggle
  - TreeItemToggleIcon
- [x] truncate (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/86a99663d5d75e2065f7ce4278d9be2effbcae04)
  - Truncate
- [x] visually-hidden (https://github.com/trendmicro-frontend/tonic-ui/pull/922/commits/9ed8c1488656fb1085bbc2f677a4bf329e0bfd19)
  - VisuallyHidden

___

### **PR Type**
enhancement, tests, documentation

___

### **Description**
- Introduced `DefaultPropsProvider` component and `useDefaultProps` hook to manage default props in React components.
- Integrated `DefaultPropsProvider` into `ThemeProvider` to pass theme components.
- Applied `useDefaultProps` to `Accordion` and `Alert` components for simplified props handling.
- Added `resolveProps` function and corresponding tests to merge default and provided props.
- Updated documentation to include examples and configuration for default props.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>7 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>DefaultPropsProvider.js</strong><dd><code>Add DefaultPropsProvider component for default props context</code></dd></summary>
<hr>

packages/react/src/default-props/DefaultPropsProvider.js

<li>Introduced <code>DefaultPropsProvider</code> component.<br> <li> Provides default props context to children components.<br>

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-4443360626424d916b7b15728905151d5f2ea853fe6b47908133968f1f9a6015">+14/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>useDefaultProps.js</strong><dd><code>Implement useDefaultProps hook for resolving props</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/react/src/default-props/useDefaultProps.js

<li>Implemented <code>useDefaultProps</code> hook.<br> <li> Resolves component props with default props from context.<br>

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-9e2b08718529118de37fa2be5ac0db9b273da8acad370e183c598387a0da497f">+41/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resolveProps.js</strong><dd><code>Add resolveProps function for merging props</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/react/src/default-props/resolveProps.js

<li>Added <code>resolveProps</code> function.<br> <li> Merges default props with provided props.<br>

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-53301b65a7bef0cc35b47b0f8ba99680614b01378a1e0da0688800746fd01095">+48/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ThemeProvider.js</strong><dd><code>Integrate DefaultPropsProvider into ThemeProvider</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/react/src/theme/ThemeProvider.js

<li>Integrated <code>DefaultPropsProvider</code> into <code>ThemeProvider</code>.<br> <li> Passes theme components to <code>DefaultPropsProvider</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-160beeffe0ffadbdf65362186091704b4ba343fcce76c7914d9954625048c064">+8/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Accordion.js</strong><dd><code>Apply useDefaultProps to Accordion component</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/react/src/accordion/Accordion.js

<li>Applied <code>useDefaultProps</code> to <code>Accordion</code> component.<br> <li> Simplified props handling with default props.<br>

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-cfa647c29e6203dbe5c3fb6329e955573b629bf5e0339450a74841e163a8e8ed">+4/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Alert.js</strong><dd><code>Apply useDefaultProps to Alert component</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/react/src/alert/Alert.js

<li>Applied <code>useDefaultProps</code> to <code>Alert</code> component.<br> <li> Simplified props handling with default props.<br>

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-d8cfcfaae3e2bcf738bd9ccc98126719796b95d3f4afe3932a47de6e61b2e21f">+4/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.js</strong><dd><code>Export default props utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/react/src/index.js

<li>Exported <code>DefaultPropsProvider</code> and <code>useDefaultProps</code>.<br> <li> Made default props utilities available for import.<br>

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-c4c3158cfb4d39b9c02a04153dcace5a54c332a56108e62f2b5fe01b9ee087be">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>resolveProps.test.js</strong><dd><code>Add tests for resolveProps function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/react/src/default-props/__tests__/resolveProps.test.js

<li>Added tests for <code>resolveProps</code> function.<br> <li> Validates merging logic of default and provided props.<br>

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-7617e3a17d9989c2165f8f3246cc7aca87ed516ce079ef0db72a611ab6340e62">+105/-0</a>&nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>_app.page.js</strong><dd><code>Configure theme components for default props</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/react-docs/pages/_app.page.js

<li>Added theme components configuration for default props.<br> <li> Provided example for setting default props.<br>

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-49f34385f9684300c85addd1b7dc358db1cfc469e159754c0ea81c147fa99672">+15/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.page.mdx</strong><dd><code>Update migration guide link for icons</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/react-docs/pages/getting-started/icons/index.page.mdx

- Updated migration guide link for icons.

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/922/files#diff-f55dd194c18094e739418caa88fc922f3c5e639e802f991a79e1fa84257dafda">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by cheton** [comment]
./review

**Comment by cheton** [comment]
/ask I want to use a JSX parser to scan files and verify if the name in `useDefaultProps` matches the component name.

Example:
```jsx
const AccordionBody = forwardRef((inProps, ref) => {
  const props = useDefaultProps({ props: inProps, name: 'AccordionBody' });
```

**Comment by cheton** [comment]
/ask I want to use a JSX parser to scan files and verify if the name in `useDefaultProps` matches the component name. I want to traverse files that only imports the "useDefaultProps" hook.

Example:
```jsx
const AccordionBody = forwardRef((inProps, ref) => {
  const props = useDefaultProps({ props: inProps, name: 'AccordionBody' });
```

Here is my code that requires further refinements:
```js
const fs = require('node:fs');
const { globSync } = require('glob');
const parser = require('@babel/parser');
const traverse = require('@babel/traverse').default;

const sourceDirectory = './';

const files = globSync([
  '*/**/*.js',
], {
  'ignore': [
    '*/**/__tests__/**/*.js',
  ],
}).sort();

for (const file of files) {
  const code = fs.readFileSync(file, { encoding: 'utf-8' });
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['jsx'],
  });

  traverse(ast, {
    VariableDeclarator(path) {
      if (path.node.init && path.node.init.callee && path.node.init.callee.name === 'forwardRef') {
        const componentName = path.node.id.name;
        path.traverse({
          CallExpression(innerPath) {
            if (innerPath.node.callee.name === 'useDefaultProps') {
              const nameProperty = innerPath.node.arguments[0].properties.find(prop => prop.key.name === 'name');
              if (nameProperty && nameProperty.value.value !== componentName) {
                console.log(`Mismatch found: ${nameProperty.value.value} does not match ${componentName}`);
              } else {
                //console.log(`Found: ${nameProperty.value.value} match ${componentName}`);
              }
            }
          }
        });
      }
    }
  });
}

```

**Comment by cheton** [comment]
/ask I want to use a JSX parser to scan files and verify if the name in `useDefaultProps` matches the component name. I want to traverse files that only imports the "useDefaultProps" hook.

Example:
```jsx
const AccordionBody = forwardRef((inProps, ref) => {
  const props = useDefaultProps({ props: inProps, name: 'AccordionBody' });
```

Here is my code for reference:
```js
const fs = require('node:fs');
const { globSync } = require('glob');
const parser = require('@babel/parser');
const traverse = require('@babel/traverse').default;

const sourceDirectory = './';

const files = globSync([
  '*/**/*.js',
], {
  'ignore': [
    '*/**/__tests__/**/*.js',
  ],
}).sort();

for (const file of files) {
  const code = fs.readFileSync(file, { encoding: 'utf-8' });
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['jsx'],
  });

  traverse(ast, {
    VariableDeclarator(path) {
      if (path.node.init && path.node.init.callee && path.node.init.callee.name === 'forwardRef') {
        const componentName = path.node.id.name;
        path.traverse({
          CallExpression(innerPath) {
            if (innerPath.node.callee.name === 'useDefaultProps') {
              const nameProperty = innerPath.node.arguments[0].properties.find(prop => prop.key.name === 'name');
              if (nameProperty && nameProperty.value.value !== componentName) {
                console.log(`Mismatch found: ${nameProperty.value.value} does not match ${componentName}`);
              } else {
                //console.log(`Found: ${nameProperty.value.value} match ${componentName}`);
              }
            }
          }
        });
      }
    }
  });
}

```

**Comment by cheton** [comment]
/ask I need to use a JSX parser to scan files and check if the name in `useDefaultProps` matches the component name. The parser should only traverse files that import the `useDefaultProps` hook. Keep in mind, some files may not include `forwardRef`, so please adjust my code so it doesn't depend on `forwardRef`.

Example:
```jsx
const AccordionBody = forwardRef((inProps, ref) => {
  const props = useDefaultProps({ props: inProps, name: 'AccordionBody' });
```

Here is my code for reference:
```js
const fs = require('node:fs');
const { globSync } = require('glob');
const parser = require('@babel/parser');
const traverse = require('@babel/traverse').default;

const sourceDirectory = './';

const files = globSync([
  '*/**/*.js',
], {
  'ignore': [
    '*/**/__tests__/**/*.js',
  ],
}).sort();

for (const file of files) {
  const code = fs.readFileSync(file, { encoding: 'utf-8' });
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['jsx'],
  });

  traverse(ast, {
    VariableDeclarator(path) {
      if (path.node.init && path.node.init.callee && path.node.init.callee.name === 'forwardRef') {
        const componentName = path.node.id.name;
        path.traverse({
          CallExpression(innerPath) {
            if (innerPath.node.callee.name === 'useDefaultProps') {
              const nameProperty = innerPath.node.arguments[0].properties.find(prop => prop.key.name === 'name');
              if (nameProperty && nameProperty.value.value !== componentName) {
                console.log(`Mismatch found: ${nameProperty.value.value} does not match ${componentName}`);
              } else {
                //console.log(`Found: ${nameProperty.value.value} match ${componentName}`);
              }
            }
          }
        });
      }
    }
  });
}

```

**Comment by cheton** [comment]
/ask I need to use a JSX parser to scan files and check if the name in `useDefaultProps` matches the component name. The parser should only traverse files that import the `useDefaultProps` hook. Keep in mind, some files may not include `forwardRef`, so please adjust my code so it doesn't depend on `forwardRef`.

Example:
```jsx
const AccordionBody = forwardRef((inProps, ref) => {
  const props = useDefaultProps({ props: inProps, name: 'AccordionBody' });
```

Here is my code for reference:
```js
const fs = require('node:fs');
const { globSync } = require('glob');
const parser = require('@babel/parser');
const traverse = require('@babel/traverse').default;

const sourceDirectory = './';

const files = globSync([
  '*/**/*.js',
], {
  'ignore': [
    '*/**/__tests__/**/*.js',
  ],
}).sort();

for (const file of files) {
  const code = fs.readFileSync(file, { encoding: 'utf-8' });
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['jsx'],
  });

  traverse(ast, {
    VariableDeclarator(path) {
      if (path.node.init && path.node.init.callee && path.node.init.callee.name === 'forwardRef') {
        const componentName = path.node.id.name;
        path.traverse({
          CallExpression(innerPath) {
            if (innerPath.node.callee.name === 'useDefaultProps') {
              const nameProperty = innerPath.node.arguments[0].properties.find(prop => prop.key.name === 'name');
              if (nameProperty && nameProperty.value.value !== componentName) {
                console.log(`Mismatch found: ${nameProperty.value.value} does not match ${componentName}`);
              } else {
                //console.log(`Found: ${nameProperty.value.value} match ${componentName}`);
              }
            }
          }
        });
      }
    }
  });
}

```

File: packages/react/src/checkbox/CheckboxControlBox.js [hunk 1]
```diff
@@ -11,16 +11,14 @@ import {
...
+  const {
     indeterminate,
     size = defaultSize,
     variantColor = defaultVariantColor,
     sx: sxProp,
     ...rest
-  },
-  ref,
-) => {
+  } = inProps;
```
Comment (tinaClin) [inline packages/react/src/checkbox/CheckboxControlBox.js:21]:
```suggestion
  } = useDefaultProps({ props: inProps, name: 'CheckboxControlBox' });
```

File: packages/react/src/radio/Radio.js [hunk 2]
```diff
@@ -2,33 +2,40 @@ import { useMergeRefs } from '@tonic-ui/react-hooks';
...
+  let checked = checkedProp;
+  let disabled = disabledProp;
+  let name = nameProp;
+  let onChange = onChangeProp;
+  let size = sizeProp;
+  let variantColor = variantColorProp;
+
   const inputRef = useRef();
   const combinedInputRef = useMergeRefs(inputRefProp, inputRef);
   const styleProps = useRadioStyle({ disabled });
```
Comment (tinaClin) [inline packages/react/src/radio/Radio.js]: Please move `const styleProps = useRadioStyle({ disabled });` to line 71.

File: packages/react/src/radio/RadioControlBox.js [hunk 3]
```diff
@@ -9,15 +9,13 @@ import {
...
+const RadioControlBox = forwardRef((inProps, ref) => {
+  const {
     size = defaultSize,
     variantColor = defaultVariantColor,
     sx: sxProp,
     ...rest
-  },
-  ref,
-) => {
+  } = inProps;
```
Comment (tinaClin) [inline packages/react/src/radio/RadioControlBox.js:18]:
```suggestion
  } = useDefaultProps({ props: inProps, name: 'RadioControlBox' });
```

File: packages/react/src/switch/SwitchControlBox.js [hunk 4]
```diff
@@ -7,15 +7,13 @@ import { defaultSize, defaultVariantColor } from './constants';
...
+const SwitchControlBox = forwardRef((inProps, ref) => {
+  const {
     size = defaultSize,
     variantColor = defaultVariantColor,
     sx: sxProp,
     ...rest
-  },
-  ref,
-) => {
+  } = inProps;
```
Comment (tinaClin) [inline packages/react/src/switch/SwitchControlBox.js:16]:
```suggestion
  } = useDefaultProps({ props: inProps, name: 'SwitchControlBox' });
```

File: packages/react/src/radio/Radio.js (hunk 2 above)
Comment (cheton) [reply packages/react/src/radio/Radio.js]: resolved in b7878874bbdbe4cf56972f3aa322f3f14b3cbb3a

**Review by tinaClin — APPROVED** [review]

**Pull request merged**